SDWebImage/Core/SDImageIOAnimatedCoder.m

```diff
@@ -40,26 +41,31 @@ @implementation SDImageIOAnimatedCoder {
     BOOL _finished;
     BOOL _preserveAspectRatio;
     CGSize _thumbnailSize;
+    dispatch_semaphore_t _lock;
 }
 
 - (void)dealloc
 {
+    SD_LOCK(_lock);
     if (_imageSource) {
         CFRelease(_imageSource);
         _imageSource = NULL;
     }
+    SD_UNLOCK(_lock);
```

Comment:
Is it helpful to add a lock inside of `dealloc` method? How is it possible that it will be called from 2+ threads at the same time?

Reply:
Correct. Each coder's decoding will be processed guardered by lock.

If you use `SDAnimatedImageView`, it will allocate each coder to the `SDAnimatedImage`. image level.

If you use the UIImageView, all of them share a single `sharedCoder`.